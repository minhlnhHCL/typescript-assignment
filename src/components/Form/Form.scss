.custom-input {
    flex: 1;
    margin-left: 32px;
    padding: 16px;
    border-radius: 8px;
    border: 1px solid rgba(146, 139, 139, 0.723)
}

.custom-label {
    font-size: 0.875rem;
    color: rgba(92, 89, 89, 0.744);
    font-weight: 500;
    width: 200px;
}

body {
    height: 100%
}

.form-wrapper {
    margin: auto;
    height: 100%;
    max-width: 480px;
    padding: 0 16px;
    display: flex;
    align-items: center;
    justify-content: center;

    .form {
        position: relative;
        height: 300px;
        padding: 20px;
        width: 320px;
        border-radius: 8px;
        box-shadow: rgba(0, 0, 0, 0.15) 1.95px 1.95px 3.6px;
    }
}

.form-actions {
    position: absolute;
    bottom: 0px;
    left: 0px;
    width: 100%;
    padding: 18px 0px;
    background-color: rgba(243, 243, 243, 0.2);
    border-radius: 0px 0px 8px 8px;

    .action-container {
        width: 100%;
        display: flex;
        justify-content: flex-end;

        .btn {
            border-radius: 8px;
            padding: 8px;
            border-color: transparent;
            cursor: pointer;
            letter-spacing: 1.5px;
            &.cancel {
                margin-right: 8px;
                background: transparent;
            }

            &.submit {
                width: 150px;
                margin-right: 24px;
                background: rgb(103, 168, 203);
                color: #fff;

                &:disabled {
                    border: 1px solid #999999;
                    background-color: #cccccc;
                    color: #666666;
                }
            }
        }
    }
}