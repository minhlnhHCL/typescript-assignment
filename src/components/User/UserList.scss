.btn-create {
    cursor: pointer;
    padding: 8px;
    width: 100%;
    border-radius: 8px;
    border: 0;
    background: transparent;
    color: black;
    box-shadow: rgba(0, 0, 0, 0.15) 1.95px 1.95px 2.6px;
}

.link-div {
    text-decoration: none;
    color: black;
    font-size: 1.25rem;
}


.list-container {
    margin: 0 auto;
    max-width: 480px;
    padding: 0 16px;

    .list-wrapper {
        display: flex;
        flex-direction: column;
    }
}