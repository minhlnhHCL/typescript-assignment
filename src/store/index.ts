import { configureStore, PreloadedState } from '@reduxjs/toolkit'
import { useDispatch, useSelector } from 'react-redux'
import { combineReducers } from 'redux'

import users from './userSlice'
import modal from './modalSlice'
import toast from './toastSlice'

const reducer = combineReducers({
    users,
    modal,
    toast
})
const setupStore = (preloadedState?: PreloadedState<RootState>) => {
    return configureStore({
        reducer,
        preloadedState,
        middleware: (getDefaultMiddleware) =>
            getDefaultMiddleware({
                serializableCheck: false
            })
    })
}

export type RootState = ReturnType<typeof reducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']

export default setupStore
