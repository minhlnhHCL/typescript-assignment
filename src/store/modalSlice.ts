import { createSlice, PayloadAction } from "@reduxjs/toolkit"


export interface IModal {
    open: boolean,
    title: string,
    children: JSX.Element | string | null
    successCb: () => void

}

const initialState: IModal = {
    open: false,
    title: '',
    children: null,
    successCb: () => { },
}

export const modalSlice = createSlice({
    name: 'modal',
    initialState,
    reducers: {
        setModal: (state, action: PayloadAction<IModal>) => {
            state.children = action.payload.children
            state.open = action.payload.open
            state.title = action.payload.title
            state.successCb = action.payload.successCb
        }
    }
})

export const { setModal } = modalSlice.actions
export default modalSlice.reducer