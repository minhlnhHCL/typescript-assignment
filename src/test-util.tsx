import React, { PropsWithChildren, ReactNode } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import type { PreloadedState } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'

import setupStore from 'store'
import type { AppStore, RootState } from 'store/index'
import { LoadingContext, LoadingProvider } from 'context/LoadingContext'
import { BrowserRouter } from 'react-router-dom'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  onLoading: () => void,
  offLoading: () => void,
}

export * from '@testing-library/react'

export const renderWithRedux = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    onLoading = () => { },
    offLoading = () => { },
    ...renderOptions
  }: ExtendedRenderOptions = { onLoading: () => { }, offLoading: () => { } }
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return <BrowserRouter>
      <Provider store={store}>
        <LoadingContext.Provider value={{ offLoading, onLoading }}>
          {children}
        </LoadingContext.Provider>
      </Provider></BrowserRouter>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}