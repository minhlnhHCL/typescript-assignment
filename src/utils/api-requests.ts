import axios from 'axios';
import { IUser } from '../components/User/UserInterface';

export interface IApiRequest {
  url?: string | '',
  body?: IUser
}

export const baseUrl = 'https://dummyjson.com/users';

export const get = ({url}: IApiRequest) => {
  return axios.get(baseUrl + url);
};

export const post = ({ url, body }: IApiRequest) => {
  return axios.post(baseUrl + url, body);
};

export const put = ({ url, body }: IApiRequest) => {
  return axios.put(baseUrl + url, body);
};

export const deleteReq = ({ url }: IApiRequest) => {
  return axios.delete(baseUrl + url);
};

