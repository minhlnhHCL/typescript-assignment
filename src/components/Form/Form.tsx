import React, { FormEvent, useContext, useEffect, useState } from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import { addUser, editUser } from '../../store/userSlice'
import './Form.scss'
import { useAppDispatch, useAppSelector } from 'store/hooks'
import { put, post, get } from 'utils/api-requests'
import { ActionCreatorWithPayload } from '@reduxjs/toolkit'
import { IUser } from '../User/UserInterface'
import { useLoadingContext } from 'context/LoadingContext'

const Form = () => {
    //Declare initial values
    const { users } = useAppSelector(state => state.users)
    const dispatch = useAppDispatch()
    const navigate = useNavigate()
    const { userId } = useParams()
    const { onLoading, offLoading } = useLoadingContext()

    const [data, setData] = useState<IUser>({
        id: 0,
        firstName: '',
        lastName: '',
        image: '',
        age: 0,
        company: {
            title: ''
        }
    })

    useEffect(() => {
        const fetchUser = async () => {
            onLoading()
            try {
                const res = await get({ url: `/${userId}` })
                setData(res.data as IUser)
            } catch (error) {
                console.log(error)
            } finally {
                offLoading()
            }
        }
        if (userId)
            fetchUser()
    }, [])

    //Form Functions
    const successHandler = (res: any, reducer: ActionCreatorWithPayload<IUser, string>) => {
        offLoading()
        dispatch(reducer(res.data))
        navigate('/')
    }

    const onSubmit = async (event: FormEvent) => {
        event.preventDefault()
        onLoading()
        try {
            let url = '/add'
            if (userId) {
                url = `/${userId}`
                const req = await put({ url: url, body: data })
                return successHandler(req, editUser)

            }
            const newId = users.length == 0 ? 31 : 32
            data.id = newId
            data.image = 'https://robohash.org/hicveldicta.png'
            const newRes = { data: data }
            await post({ url: url, body: data })
            return successHandler(newRes, addUser)


        } catch (error) {
            offLoading()
            console.log(error)
        }
    }

    return (
        <div className='form-wrapper'>
            <form onSubmit={onSubmit} className='form'>
                <div className='d-flex-ac'>
                    <label htmlFor='firstName' className='custom-label'>First Name</label>
                    <input id='firstName' className='custom-input ' placeholder='First Name' name='firstName' value={data.firstName} onChange={(e) => setData({ ...data, firstName: e.target.value })} />
                </div>
                <div className='d-flex-ac mt-32'>
                    <label htmlFor='lastName' className='custom-label'>Last Name</label>
                    <input id='lastName' className='custom-input' placeholder='Last Name' name='lastName' value={data.lastName} onChange={(e) => setData({ ...data, lastName: e.target.value })} />
                </div>
                <div className='d-flex-ac mt-32'>
                    <label htmlFor='companyTitle' className='custom-label'>Company Title</label>
                    <input id='companyTitle' className='custom-input' placeholder='Company Title' name='companyTitle' value={data.company?.title} onChange={(e) => setData({ ...data, company: { ...data.company, title: e.target.value } })} />
                </div>

                <div className='form-actions'>
                    <div className="action-container">
                        <button className='btn cancel' type='button' onClick={() => {
                            navigate('/')
                        }}>
                            Cancel
                        </button>
                        <button aria-label='submit' name='btn-submit' className='btn submit' disabled={!(data.firstName && data.lastName && data.company?.title)} type='submit'>Submit</button>
                    </div>
                </div>
            </form >
        </div>
    )
}

export default Form