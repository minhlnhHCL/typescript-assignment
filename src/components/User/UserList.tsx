import { useEffect } from 'react'
import './UserList.scss'
import { useNavigate } from 'react-router-dom'
import { getUsers } from 'store/userSlice'
import { get } from 'utils/api-requests'
import { useAppDispatch, useAppSelector } from 'store/hooks'
import UserWrapper from './UserWrapper/UserWrapper'
import { IUser } from './UserInterface'
import { useLoadingContext } from 'context/LoadingContext'

const UserList = () => {
    const { users } = useAppSelector(state => state.users)
    const dispatch = useAppDispatch()
    const navigate = useNavigate()
    const {onLoading, offLoading} = useLoadingContext()

    useEffect(() => {
        onLoading()
        const getData = async () => {
            try {
                if (users.length === 0 || users.length === 1) {
                    const res = await get({url: ''})
                    if (res.status === 200) {
                        return dispatch(getUsers(res.data.users))
                    }
                }
            } catch (error) {
                console.log(error)
            } finally {
                offLoading()
            }
        }
        getData()
    }, [users])

    return (
        <div className='list-container mt-32'>
            <button className='btn-create' onClick={() => {
                navigate('/create')
            }}>Create New User</button>
            <div className='list-wrapper'>
                {
                    users !== undefined && users.length !== 0 && users.map((item: IUser, idx) => (
                        <UserWrapper key={`${item.id}-${idx}`} data={item} />
                    ))
                }
            </div>
        </div>
    )
}

export default UserList