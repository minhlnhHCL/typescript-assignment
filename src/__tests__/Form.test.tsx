import Form from "components/Form/Form"
import {
    cleanup,
    fireEvent,
    screen,
    waitFor
} from "@testing-library/react"
import "@testing-library/jest-dom"
import { useNavigate, useParams } from "react-router-dom"
import { renderWithRedux } from "test-util"
import { get, post, put } from "utils/api-requests"
import { useAppDispatch, useAppSelector } from "store/hooks"
import { IUser } from "components/User/UserInterface"
import reducer, { addUser, deleteUser, editUser, UserState } from 'store/userSlice'


jest.mock("store/hooks", () => ({
    ...jest.requireActual("react-redux"),
    useAppSelector: jest.fn(),
    useAppDispatch: jest.fn(),
}))

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
    useParams: jest.fn()
}))

jest.mock('utils/api-requests')

const editUserData: IUser = {
    id: 1,
    firstName: "Minh",
    lastName: "Luong",
    image: 'https://robohash.org/hicveldicta.png',
    age: 25,
    company: {
        title: "Front End Engineer",
    },
}


const addUserData: IUser = {
    firstName: "Bi",
    lastName: "Bi",
    image: 'https://robohash.org/hicveldicta.png',
    age: 0,
    company: {
        title: "Front End Engineer",
    },
}

const sampleUsers: IUser[] = [{
    id: 1,
    firstName: "Minh",
    lastName: "Luong",
    image: 'https://robohash.org/hicveldicta.png',
    age: 25,
    company: {
        title: "Front End Engineer",
    },
}, {
    id: 2,
    firstName: "Min",
    lastName: "Min",
    image: 'https://robohash.org/hicveldicta.png',
    age: 21,
    company: {
        title: "Front End Engineer",
    },
}]

describe("User Form", () => {
    beforeEach(() => {
        (useAppDispatch as jest.Mock).mockImplementationOnce(() => jest.fn());
        (useNavigate as jest.Mock).mockReturnValue(() => { });
        (useAppSelector as jest.Mock).mockImplementation(() => { });
    })

    afterAll(cleanup);

    it("should display an empty form", async () => {
        (useParams as jest.Mock).mockReturnValue({ userId: null });
        (useAppSelector as jest.Mock).mockReturnValue({
            users: []
        });
        const { container } = renderWithRedux(<Form />);
        expect(container).toMatchSnapshot();
    })

    it("should add a new user", async () => {
        (useParams as jest.Mock).mockReturnValue({ userId: null });
        (useAppSelector as jest.Mock).mockReturnValue({
            users: [],
        });

        const onLoading = jest.fn();
        const offLoading = jest.fn();
        const navigate = jest.fn();
        const dispatch = jest.fn();
        const newInput = {
            ...addUserData,
            id: 31
        };
        (useAppDispatch as jest.Mock).mockReturnValue(dispatch);
        (useNavigate as jest.Mock).mockReturnValue(navigate);
        (post as jest.Mock).mockResolvedValue({
            data: newInput
        });

        renderWithRedux(<Form />, { onLoading, offLoading });

        fireEvent.change(screen.getByLabelText('First Name'), {
            target: { value: newInput.firstName, name: 'firstName' }
        });

        fireEvent.change(screen.getByLabelText('Last Name'), {
            target: { value: newInput.lastName, name: 'lastName' }
        });

        fireEvent.change(screen.getByLabelText('Company Title'), {
            target: { value: newInput.company?.title, name: 'companyTitle' }
        });

        expect(screen.getByRole('button', { name: 'submit' }));
        // screen.getByRole('button', { name: 'submit' });
        fireEvent.click(screen.getByRole('button', { name: 'submit' }));
        await waitFor(() => expect(onLoading).toBeCalled());
        expect(post).toBeCalled();
        expect(post).toBeCalledWith({ url: '/add', body: newInput });
        await waitFor(() => expect(offLoading).toBeCalled());
        expect(dispatch).toBeCalled();
        // expect(dispatch).toBeCalledWith(reducer({ users: [] }, addUser(newInput)));
        expect(navigate).toBeCalled();
        expect(navigate).toBeCalledWith('/');

    });

    it("should submit edit user", async () => {
        (useParams as jest.Mock).mockReturnValue({ userId: '1' });
        (useAppSelector as jest.Mock).mockReturnValue({
            users: sampleUsers
        });

        const onLoading = jest.fn();
        const offLoading = jest.fn();
        const navigate = jest.fn();
        const newInput = {
            ...editUserData,
            firstName: 'Ter',
            lastName: 'Med',
            company: {
                title: 'IT help desk'
            }
        };

        (get as jest.Mock).mockResolvedValueOnce({
            data: editUserData
        });

        (put as jest.Mock).mockResolvedValue({
            data: newInput
        });

        (useNavigate as jest.Mock).mockReturnValue(navigate);


        const { container } = renderWithRedux(<Form />, { onLoading, offLoading });
        await waitFor(() => expect(onLoading).toBeCalled());
        expect(get).toBeCalled();
        expect(get).toBeCalledWith({ url: `/${editUserData.id}` });
        await waitFor(() => expect(offLoading).toBeCalled());
        fireEvent.change(screen.getByLabelText('First Name'), {
            target: { value: newInput.firstName, name: 'firstName' }
        });

        fireEvent.change(screen.getByLabelText('Last Name'), {
            target: { value: newInput.lastName, name: 'lastName' }
        });

        fireEvent.change(screen.getByLabelText('Company Title'), {
            target: { value: newInput.company.title, name: 'companyTitle' }
        });

        fireEvent.click(screen.getByRole('button', { name: 'submit' }));
        await waitFor(() => expect(onLoading).toBeCalled());
        expect(put).toBeCalled();
        expect(put).toBeCalledWith({ url: `/${newInput.id}`, body: newInput });
        expect(useAppDispatch).toBeCalled();
        await waitFor(() => expect(offLoading).toBeCalled());
        expect(navigate).toBeCalled();
        expect(navigate).toBeCalledWith('/');
    })
})
