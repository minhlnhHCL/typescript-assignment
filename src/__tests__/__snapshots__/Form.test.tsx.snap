// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Form should display an empty form 1`] = `
<div>
  <div
    class="form-wrapper"
  >
    <form
      class="form"
    >
      <div
        class="d-flex-ac"
      >
        <label
          class="custom-label"
          for="firstName"
        >
          First Name
        </label>
        <input
          class="custom-input "
          id="firstName"
          name="firstName"
          placeholder="First Name"
          value=""
        />
      </div>
      <div
        class="d-flex-ac mt-32"
      >
        <label
          class="custom-label"
          for="lastName"
        >
          Last Name
        </label>
        <input
          class="custom-input"
          id="lastName"
          name="lastName"
          placeholder="Last Name"
          value=""
        />
      </div>
      <div
        class="d-flex-ac mt-32"
      >
        <label
          class="custom-label"
          for="companyTitle"
        >
          Company Title
        </label>
        <input
          class="custom-input"
          id="companyTitle"
          name="companyTitle"
          placeholder="Company Title"
          value=""
        />
      </div>
      <div
        class="form-actions"
      >
        <div
          class="action-container"
        >
          <button
            class="btn cancel"
            type="button"
          >
            Cancel
          </button>
          <button
            aria-label="submit"
            class="btn submit"
            disabled=""
            name="btn-submit"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
